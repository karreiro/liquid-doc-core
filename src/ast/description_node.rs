use serde::{Deserialize, Serialize};

use super::{position::Position, TextNode};

//   {
//     "type": "LiquidDocDescriptionNode",
//     "name": "description",
//     "position": {
//       "start": 10,
//       "end": 16
//     },
//     "source": "{% doc %}\nkdkd\n\n@param {sometype} requiredParamWithSomeType - This is a cool parameter\n{% enddoc %}",
//     "content": {
//       "type": "TextNode",
//       "value": "kdkd\n\n",
//       "position": {
//         "start": 10,
//         "end": 16
//       },
//       "source": "{% doc %}\nkdkd\n\n@param {sometype} requiredParamWithSomeType - This is a cool parameter\n{% enddoc %}"
//     },
//     "isImplicit": true,
//     "isInline": true
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct LiquidDocDescriptionNode {
    pub content: TextNode,
    #[serde(rename = "isImplicit")]
    pub is_implicit: bool,
    #[serde(rename = "isInline")]
    pub is_inline: bool,
    pub position: Position,
    pub source: String,
    pub name: String,
}

impl LiquidDocDescriptionNode {
    fn new(
        content: TextNode,
        is_implicit: bool,
        is_inline: bool,
        position: Position,
        source: String,
    ) -> Self {
        LiquidDocDescriptionNode {
            content,
            is_implicit,
            is_inline,
            position,
            source,
            name: "description".to_string(), // The node name is always "description"
        }
    }
    pub fn explicit(
        pair: &pest::iterators::Pair<crate::parser::Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        assert!(
            pair.as_rule() == crate::parser::Rule::descriptionNode,
            "Expected a descriptionNode, found {:?}",
            pair.as_rule()
        );

        let content = TextNode::from_pair(pair, position_offset);
        let source_str = pair.as_str();

        LiquidDocDescriptionNode::new(
            content,
            false,
            true,
            Position::from_pair(pair, position_offset),
            source_str.to_string(),
        )
    }
    pub fn implicit(
        pair: &pest::iterators::Pair<crate::parser::Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        assert!(
            pair.as_rule() == crate::parser::Rule::ImplicitDescription,
            "Expected a ImplicitDescription, found {:?}",
            pair.as_rule()
        );

        let content_node = pair.clone().into_inner().next().unwrap();

        let content = TextNode::from_pair(&content_node, position_offset);
        let source_str = pair.as_str();

        LiquidDocDescriptionNode::new(
            content,
            true,
            true,
            Position::from_pair(pair, position_offset),
            source_str.to_string(),
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::{ast::LiquidNode, parser::parse_liquid_string};

    #[test]
    fn test_liquid_doc_implicit_description_node() {
        let input =
            "kdkd\n\n@param {sometype} requiredParamWithSomeType - This is a cool parameter";
        let ast = parse_liquid_string(input, None).unwrap();
        let description_node = match ast.nodes.first() {
            Some(LiquidNode::LiquidDocDescriptionNode(node)) => node,
            _ => panic!("Expected a LiquidDocDescriptionNode"),
        };

        assert_eq!(description_node.name, "description");
        assert!(description_node.is_implicit);
        assert!(description_node.is_inline);
        assert_eq!(description_node.content.value, "kdkd\n\n");
    }
    #[test]
    fn test_liquid_doc_explicit_description_node() {
        let input = "@description kdkd
            @param {sometype} requiredParamWithSomeType - This is a cool parameter";
        let ast = parse_liquid_string(input, None).unwrap();
        let description_node = match ast.nodes.first() {
            Some(LiquidNode::LiquidDocDescriptionNode(node)) => node,
            _ => panic!("Expected a LiquidDocDescriptionNode"),
        };

        assert_eq!(description_node.name, "description");
        assert!(!description_node.is_implicit);
        assert!(description_node.is_inline);
        assert_eq!(description_node.content.value, "@description kdkd\n");
    }
}
