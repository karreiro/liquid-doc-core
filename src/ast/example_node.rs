use serde::{Deserialize, Serialize};

use super::{position::Position, TextNode};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct LiquidDocExampleNode {
    pub content: TextNode,
    #[serde(rename = "isInline")]
    pub is_inline: bool,
    pub position: Position,
    pub source: String,
    pub name: String,
}

impl LiquidDocExampleNode {
    pub fn new(content: TextNode, is_inline: bool, position: Position, source: String) -> Self {
        LiquidDocExampleNode {
            content,
            is_inline,
            position,
            source,
            name: "example".to_string(), // The node name is always "example"
        }
    }

    pub fn from_pair(
        pair: &pest::iterators::Pair<crate::parser::Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        assert!(
            pair.as_rule() == crate::parser::Rule::exampleNode,
            "Expected a exampleNode, found {:?}",
            pair.as_rule()
        );

        let mut content = TextNode::from_pair(pair, position_offset);
        content.value = content.value.trim_start_matches("@example ").to_string();
        content.position.shift_start(9); // Adjust for the "@example " prefix

        let position = Position::from_pair(pair, position_offset);
        let source = pair.as_str().to_string();
        LiquidDocExampleNode::new(content, true, position, source)
    }
}

#[cfg(test)]
mod test {
    use crate::{ast::LiquidNode, parser::parse_liquid_string};
    use pretty_assertions::assert_eq;

    #[test]
    fn test_parse_optional_param_with_type_and_description() {
        let input = "@example simple inline example\n";
        let result = parse_liquid_string(input, Some(10));

        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocExampleNode(example_node) = node {
            assert_eq!(example_node.content.as_str(), "simple inline example\n");
            assert!(example_node.is_inline);
        } else {
            panic!("Expected a LiquidDocExampleNode");
        }
    }

    #[test]
    pub fn test_serialization_round_trip() {
        let input = "@example simple inline example\n";
        let ast = parse_liquid_string(input, Some(10)).unwrap();

        let expected = r#"[
  {
    "type": "LiquidDocExampleNode",
    "content": {
      "value": "simple inline example\n",
      "position": {
        "start": 19,
        "end": 41
      },
      "source": "{% doc %}\n@example simple inline example\n{% enddoc %}",
      "type": "TextNode"
    },
    "isInline": true,
    "position": {
      "start": 10,
      "end": 41
    },
    "source": "{% doc %}\n@example simple inline example\n{% enddoc %}",
    "name": "example"
  }
]"#;

        let actual = serde_json::to_string_pretty(&ast.nodes).unwrap();

        assert_eq!(expected, actual);
    }
}
