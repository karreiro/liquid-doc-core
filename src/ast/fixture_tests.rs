// In your test file
include!(concat!(env!("OUT_DIR"), "/fixtures_macro.rs"));

macro_rules! generate_fixture_test {
    ($fixture_name:expr) => {
        paste::paste! {
            #[test]
            fn [<test_fixture_ $fixture_name>]() {
                let content = include_str!(concat!("../../web/fixtures/", $fixture_name, ".liquid"));
                // Your test logic here
                crate::assert_json_output!(content);
            }
        }
    };
}

// This will generate all the tests
for_each_fixture_file!(generate_fixture_test);
