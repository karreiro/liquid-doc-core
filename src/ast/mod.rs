mod description_node;
mod doc_param_node;
mod example_node;
mod liquid_ast;
mod position;
pub(crate) mod test_utils;
mod text_node;

pub use description_node::*;
pub use doc_param_node::*;
pub use example_node::*;
pub use liquid_ast::*;
pub use text_node::*;
