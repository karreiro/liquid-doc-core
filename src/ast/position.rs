use serde::{Deserialize, Serialize};

use crate::parser::Rule;

#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
pub struct Position {
    pub start: usize,
    pub end: usize,
}

impl Position {
    pub fn new(start: usize, end: usize, offset: Option<usize>) -> Self {
        match offset {
            Some(o) => Position {
                start: start + o,
                end: end + o,
            },
            None => Position { start, end },
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, offset: Option<usize>) -> Self {
        let span = pair.as_span();
        Position::new(span.start(), span.end(), offset)
    }

    pub fn shift_start(&mut self, offset: usize) {
        self.start += offset;
    }
}
