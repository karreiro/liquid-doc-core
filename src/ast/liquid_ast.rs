use crate::ast::{LiquidDocParamNameNode, LiquidDocParamNode, LiquidRawTagNode, TextNode};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum LiquidNode {
    LiquidRawTag(LiquidRawTagNode),
    TextNode(TextNode),
    LiquidDocParamNode(LiquidDocParamNode),
    LiquidDocParamNameNode(LiquidDocParamNameNode),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LiquidAST {
    pub nodes: Vec<LiquidNode>,
}

impl LiquidAST {
    pub fn new() -> Self {
        LiquidAST { nodes: Vec::new() }
    }

    pub fn add_node(&mut self, node: LiquidNode) {
        self.nodes.push(node);
    }

    pub fn head(&self) -> LiquidNode {
        self.nodes[0].clone()
    }

    pub fn len(&self) -> usize {
        self.nodes.len()
    }
}
