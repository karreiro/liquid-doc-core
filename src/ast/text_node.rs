use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::position::Position;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct TextNode {
    pub value: String,
    pub position: Position,
    pub source: String,
    #[serde(rename = "type", skip_deserializing)]
    pub type_: &'static str,
}

impl TextNode {
    pub fn new(value: String, position: Position, source: String) -> Self {
        TextNode {
            value,
            position,
            source,
            type_: "TextNode",
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, position_offset: Option<usize>) -> Self {
        let source_str = pair.as_str();
        TextNode::new(
            source_str.to_string(),
            Position::from_pair(pair, position_offset),
            source_str.to_string(),
        )
    }

    pub fn without_brackets(
        pair: &pest::iterators::Pair<Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        let mut text_node = Self::from_pair(pair, position_offset);
        let new_value = text_node
            .value
            .trim_matches(|c| c == '{' || c == '}' || c == '[' || c == ']')
            .to_string();
        if new_value == text_node.value {
            return text_node; // No change needed
        }
        text_node.value = new_value;

        text_node.position.shift_start(1); // Adjust position to account for removed brackets
        text_node.position.shift_end_down(1); // Adjust end position as well
        text_node
    }

    pub fn trim_content_start(&mut self, to_strip: &str) {
        if self.value.starts_with(to_strip) {
            self.value = self.value.trim_start_matches(to_strip).to_string();
            self.position.shift_start(to_strip.len());
        }
    }

    #[allow(dead_code)]
    pub fn as_str(&self) -> &str {
        &self.value
    }
    pub fn is_empty(&self) -> bool {
        self.value.is_empty()
    }
}
