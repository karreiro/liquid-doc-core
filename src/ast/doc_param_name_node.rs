use serde::{Deserialize, Serialize};

use crate::{ast::position::Position, parser::Rule};

use super::text_node::TextNode;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LiquidDocParamNameNode {
    pub content: TextNode,
    pub position: Position,
    pub source: String,
    pub required: bool,
}

impl LiquidDocParamNameNode {
    pub fn new(pair: &pest::iterators::Pair<Rule>) -> Self {
        assert!(
            pair.as_rule() == Rule::paramName,
            "Expected a paramName, found {:?}",
            pair.as_rule()
        );

        let source_str = pair.as_str();

        LiquidDocParamNameNode {
            content: TextNode::new(pair),
            position: Position::from_pair(pair),
            source: source_str.to_string(),
            required: !source_str.starts_with('[') && !source_str.ends_with(']'),
        }
    }
}
