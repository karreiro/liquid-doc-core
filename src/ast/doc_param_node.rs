use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::doc_param_name_node::LiquidDocParamNameNode;
use super::text_node::TextNode;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LiquidDocParamNode {
    pub name: String,
    #[serde(rename = "locStart")]
    pub loc_start: usize,
    #[serde(rename = "locEnd")]
    pub loc_end: usize,
    pub source: String,
    #[serde(rename = "paramType")]
    pub param_type: Option<TextNode>,
    #[serde(rename = "paramName")]
    pub param_name: LiquidDocParamNameNode,
    #[serde(rename = "paramDescription")]
    pub param_description: Option<TextNode>,
}
impl LiquidDocParamNode {
    pub fn new(pair: &pest::iterators::Pair<Rule>) -> Self {
        println!("{:#?}", pair);
        assert!(
            pair.as_rule() == Rule::paramNode,
            "Expected a paramNode, found {:?}",
            pair.as_rule()
        );

        let mut inner = pair.clone().into_inner();

        let first = inner.next().expect("Expected at least one inner pair");

        let (param_type, name) = if let Rule::paramType = first.as_rule() {
            (
                // Remove the curly braces from the type string
                Some(TextNode::new_trim_ends(&first)),
                inner.next().expect("Expected a paramName after paramType"),
            )
        } else {
            (None, first)
        };

        let param_name = LiquidDocParamNameNode::new(&name);

        let description = inner.next().and_then(|t| {
            if !t.as_str().is_empty() {
                Some(TextNode::new(&t))
            } else {
                None
            }
        });

        // let pair = pair.next
        let source_str = pair.as_str();
        let span = pair.as_span();
        LiquidDocParamNode {
            name: "param".to_string(), // The node name is always "param"
            loc_start: span.start(),
            loc_end: span.end(),
            source: source_str.to_string(),
            param_type, // Default to None, can be set later
            param_name,
            param_description: description,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::ast::LiquidNode;
    use crate::parser::parse_liquid_string;

    /*
        """@param requiredParamWithNoType
            @param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.
            @param {String} paramWithNoDescription
            @param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description
            @param [optionalParameterWithDescription] - optional parameter description
            @param {String} [optionalParameterWithType]
            @unsupported this node falls back to a text node
    """
         */
    #[test]
    fn test_parse_param_with_type() {
        let input = "@param {sometype} requiredParamWithNoType";
        let result = parse_liquid_string(input);

        println!("{:#?}", result);
        assert!(result.is_some());
        let node = result.unwrap().head();

        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.content.as_str(),
                "requiredParamWithNoType"
            );
            assert!(param_node.param_description.is_none());
            assert!(param_node.param_type.is_some());
            assert_eq!(param_node.param_type.unwrap().as_str(), "sometype");
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }
    #[test]
    fn test_parse_param_with_type_and_description() {
        let input = "@param {sometype} requiredParamWithNoType - This is a cool parameter";
        let result = parse_liquid_string(input);

        println!("{:#?}", result);
        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.content.as_str(),
                "requiredParamWithNoType"
            );
            assert_eq!(
                param_node.param_description.unwrap().as_str(),
                "This is a cool parameter"
            );
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }
}
