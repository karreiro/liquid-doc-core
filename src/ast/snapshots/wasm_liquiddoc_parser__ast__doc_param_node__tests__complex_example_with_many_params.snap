---
source: src/ast/doc_param_node.rs
expression: "@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]"
---
[
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 10,
      "end": 40
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "requiredParamWithNoType",
      "position": {
        "start": 17,
        "end": 40
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": null,
    "paramType": null,
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 41,
      "end": 161
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "paramWithDescription",
      "position": {
        "start": 57,
        "end": 77
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "param with description and `punctation`. This is still a valid param description.",
      "position": {
        "start": 80,
        "end": 161
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 49,
        "end": 55
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 162,
      "end": 200
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "paramWithNoDescription",
      "position": {
        "start": 178,
        "end": 200
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": null,
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 170,
        "end": 176
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 201,
      "end": 305
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithTypeAndDescription",
      "position": {
        "start": 218,
        "end": 257
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "optional parameter with type and description",
      "position": {
        "start": 261,
        "end": 305
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 209,
        "end": 215
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "required": false
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 306,
      "end": 380
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithDescription",
      "position": {
        "start": 314,
        "end": 346
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "optional parameter description",
      "position": {
        "start": 350,
        "end": 380
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramType": null,
    "required": false
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 381,
      "end": 424
    },
    "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithType",
      "position": {
        "start": 398,
        "end": 423
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "paramDescription": null,
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 389,
        "end": 395
      },
      "source": "{% doc %}\n@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]\n{% enddoc %}"
    },
    "required": false
  }
]
